src/ast/annot.ml

```diff
@@ -0,0 +1,55 @@
+open Format
+
+type annot =
+  { annotid : string
+  ; items : item list
+  }
+
+and item =
+  | Atom of string
+  | String of string
+  | Id of string
+  | Int of string
+  | Float of string
+  | Parens of item list
+  | Annot of annot
+
+module NameMap = Map.Make (String)
+
+(* TODO: find a better way to format annotations, possibly by
+      - recording extra format information when parsing
+      - defining rules specific to each sort of annotations *)
+
+let rec pp_annot fmt annot =
+  pp fmt "(annot %a@\n  @[<b 2>%a@]@\n)" pp_string annot.annotid
+    (pp_list ~pp_sep:pp_space pp_item)
+    annot.items
+
+and pp_item fmt = function
+  | Atom atom -> pp_string fmt atom
+  | String str -> pp_string fmt str
+  | Id id -> Types.pp_id fmt id
+  | Int i -> pp_string fmt i
+  | Float f -> pp_string fmt f
+  | Parens items -> pp_list ~pp_sep:pp_space pp_item fmt items
+  | Annot annot -> pp_annot fmt annot
+
+let annot_recorder : annot list NameMap.t ref = ref NameMap.empty
+
+let find_nil key map = Option.value (NameMap.find_opt key map) ~default:[]
+
+let record_annot annot =
+  let old = find_nil annot.annotid !annot_recorder in
+  annot_recorder := NameMap.add annot.annotid (annot :: old) !annot_recorder
```

Comment:
If you use a `Hashtbl` instead of a `Map` you can avoid having to record previous bindings by yourself as they'll be stored in the table by default when using `Hashtbl.add`, you can recover all bindings with `Hashtbl.find_all`.

Reply:
I believe Hashtbl to be better. :)